scripts/create-darwin-volume.sh

```diff
@@ -41,129 +243,517 @@ create_synthetic_objects(){
+<plist version="1.0">
+<dict>
+  <key>RunAtLoad</key>
+  <true/>
+  <key>Label</key>
+  <string>org.nixos.darwin-store</string>
+  <key>ProgramArguments</key>
+  <array>
+$(generate_mount_command "$1" "$2")
+  </array>
+</dict>
+</plist>
+EOF
+}
+
+_eat_bootout_err() {
+    /usr/bin/grep -v "Boot-out failed: 36: Operation now in progress" 1>&2
+}
+
+# TODO: remove with --uninstall?
+uninstall_launch_daemon_directions() {
+    substep "Uninstall LaunchDaemon $1" \
+      "  sudo launchctl bootout system/$1" \
+      "  sudo rm $2"
+}
+
+uninstall_launch_daemon_prompt() {
+    cat <<EOF
+
+The installer adds a LaunchDaemon to $3: $1
+EOF
+    if ui_confirm "Can I remove it?"; then
+        _sudo "to terminate the daemon" \
+            launchctl bootout "system/$1" 2> >(_eat_bootout_err) || true
+            # this can "fail" with a message like:
+            # Boot-out failed: 36: Operation now in progress
+        _sudo "to remove the daemon definition" rm "$2"
+    fi
+}
+
+nix_volume_mountd_uninstall_directions() {
+    uninstall_launch_daemon_directions "org.nixos.darwin-store" \
+        "$NIX_VOLUME_MOUNTD_DEST"
+}
+
+nix_volume_mountd_uninstall_prompt() {
+    uninstall_launch_daemon_prompt "org.nixos.darwin-store" \
+        "$NIX_VOLUME_MOUNTD_DEST" \
+        "mount your Nix volume"
+}
+
+# TODO: move nix_daemon to install-darwin-multi-user if/when uninstall_launch_daemon_prompt moves up to install-multi-user
+nix_daemon_uninstall_prompt() {
+    uninstall_launch_daemon_prompt "org.nixos.nix-daemon" \
+        "$NIX_DAEMON_DEST" \
+        "run the nix-daemon"
+}
+
+# TODO: remove with --uninstall?
+nix_daemon_uninstall_directions() {
+    uninstall_launch_daemon_directions "org.nixos.nix-daemon" \
+        "$NIX_DAEMON_DEST"
+}
+
+
+# TODO: remove with --uninstall?
+synthetic_conf_uninstall_directions() {
+    # :1 to strip leading slash
+    substep "Remove ${NIX_ROOT:1} from /etc/synthetic.conf" \
+        "  If nix is the only entry: sudo rm /etc/synthetic.conf" \
+        "  Otherwise: grep -v '^${NIX_ROOT:1}$' /etc/synthetic.conf | sudo dd of=/etc/synthetic.conf"
+}
+
+synthetic_conf_uninstall_prompt() {
+    cat <<EOF
+During install, I add '${NIX_ROOT:1}' to /etc/synthetic.conf, which instructs
+macOS to create an empty root directory for mounting the Nix volume.
+EOF
+    # there are a few things we can do here
+    # 1. if grep -v didn't match anything (also, if there's no diff), we know this is moot
+    # 2. if grep -v was empty (but did match?) I think we know that we can just remove the file
+    # 3. if the edit doesn't ~empty the file, show them the diff and ask
+    if ! /usr/bin/grep -v '^'"${NIX_ROOT:1}"'$' /etc/synthetic.conf > "$SCRATCH/synthetic.conf.edit"; then
+        if confirm_rm "/etc/synthetic.conf"; then
+            return 0
+        fi
+    else
+        if /usr/bin/cmp <<<"" "$SCRATCH/synthetic.conf.edit"; then
```

Comment:
`grep -v`'s error code says whether it printed anything, not whether the regex matches. `echo foo | grep -v bar` returns 0, `echo -n | grep -v bar` returns 1. It also can return an error if the file doesn't exist, though right now we only call this method if we think the file contains our root.

Which is to say, this comparison against empty is unnecessary.

Reply:
Gah. Good catch. I think I iterated to a better pattern for this anyways. I've edited the function locally to:
```bash
synthetic_conf_uninstall_prompt() {
    cat <<EOF
During install, I add '${NIX_ROOT:1}' to /etc/synthetic.conf, which instructs
macOS to create an empty root directory for mounting the Nix volume.
EOF
    # make the edit to a copy
    /usr/bin/sed -e "/^${NIX_ROOT:1}$/d" "$1" > "$SCRATCH/synthetic.conf.edit"
    # ask to rm if this left the file empty, else edit
    if /usr/bin/diff -q <(:) <(/usr/bin/grep -v "^#" "$SCRATCH/synthetic.conf.edit") &>/dev/null; then
        if confirm_rm "/etc/synthetic.conf"; then
            return 0
        fi
    else
        if confirm_edit "$SCRATCH/synthetic.conf.edit" "/etc/synthetic.conf"; then
            return 0
        fi
    fi
    # fallback instructions
    echo "Manually remove nix from /etc/synthetic.conf"
    return 1
}
```
Does that look okay?